### ISSUE TITLE
retaining unique atom names in the .pdbqt file

### ISSUE THREAD

**Issue by alfredoq** [forlilab/Meeko#47]
hello,

when preparing a .pdbqt file I noticed that the unique atom name generated by RDKit are removed. An example below:

_from rdkit import Chem
from rdkit.Chem import AllChem_

_smiles = "CCCCO"
m = Chem.MolFromSmiles(smiles)
m3d = Chem.AddHs(m)    
AllChem.EmbedMolecule(m3d)
pdb_file = Chem.MolToPDBBlock(m3d)_

in this case _pdb_file_ contains:

_HETATM    1  C1  UNL     1      -1.515   0.808   0.093  1.00  0.00           C  
HETATM    2  C2  UNL     1      -0.734  -0.344   0.622  1.00  0.00           C  
HETATM    3  C3  UNL     1       0.290  -0.899  -0.311  1.00  0.00           C  
HETATM    4  C4  UNL     1       1.341   0.103  -0.717  1.00  0.00           C  
HETATM    5  O1  UNL     1       2.046   0.592   0.381  1.00  0.00           O  
HETATM    6  H1  UNL     1      -2.495   0.475  -0.327  1.00  0.00           H  
HETATM    7  H2  UNL     1      -1.772   1.505   0.929  1.00  0.00           H  
HETATM    8  H3  UNL     1      -1.014   1.393  -0.690  1.00  0.00           H  
HETATM    9  H4  UNL     1      -1.445  -1.161   0.864  1.00  0.00           H  
HETATM   10  H5  UNL     1      -0.203  -0.023   1.548  1.00  0.00           H  
HETATM   11  H6  UNL     1       0.838  -1.695   0.263  1.00  0.00           H  
HETATM   12  H7  UNL     1      -0.134  -1.413  -1.197  1.00  0.00           H  
HETATM   13  H8  UNL     1       2.046  -0.435  -1.397  1.00  0.00           H  
HETATM   14  H9  UNL     1       0.948   0.957  -1.266  1.00  0.00           H  
HETATM   15  H10 UNL     1       1.801   0.136   1.204  1.00  0.00           H_

with atoms having unique names. However after processing this RDKit molecule to get the corresponding .pdbqt file:

_preparator = MoleculePreparation()
preparator.prepare(m3d)
pdbqt_string = preparator.write_pdbqt_string()
pdbqt_string_ready = pdbqt_string.replace('UNL','MOL')   
ligand_file = '/home/fredy/Desktop/test.pdbqt'   
with open(ligand_file,'w') as pdbqt_file:
    pdbqt_file.write(pdbqt_string_ready)_

I get:

_REMARK SMILES CCCCO
REMARK SMILES IDX 3 1 2 2 1 3 4 4 5 5
REMARK H PARENT 5 6
REMARK Flexibility Score: inf
ROOT
ATOM      1  C   MOL     1       0.290  -0.899  -0.311  1.00  0.00     0.028 C 
ENDROOT
BRANCH   1   2
ATOM      2  C   MOL     1      -0.734  -0.344   0.622  1.00  0.00    -0.001 C 
ATOM      3  C   MOL     1      -1.515   0.808   0.093  1.00  0.00     0.004 C 
ENDBRANCH   1   2
BRANCH   1   4
ATOM      4  C   MOL     1       1.341   0.103  -0.717  1.00  0.00     0.156 C 
BRANCH   4   5
ATOM      5  O   MOL     1       2.046   0.592   0.381  1.00  0.00    -0.396 OA
ATOM      6  H   MOL     1       1.801   0.136   1.204  1.00  0.00     0.210 HD
ENDBRANCH   4   5
ENDBRANCH   1   4
TORSDOF 3_

I revised _preparator_ options looking for some keyword to retain unique atom names, however I was not able to find this feature. Is it possible to retain unique atom names in the .pdbqt file?

Thanks in advance for the support,

best

**Comment by rwxayheee** [comment]
Hello @alfredoq 

The atom names, if they're accessible via `atom.GetMonomerInfo().GetName()`, will appear in the output ligand PDBQT. It didn't work the way you wanted in your case, because the atom names were determined at the moment `Chem.MolToPDBBlock(m3d)`. They then went to the `pdb_file` block but were never assigned to rdkit mol `m3d`. 

If you do this instead: 

```py
from meeko import MoleculePreparation
m3d = Chem.MolFromPDBBlock(pdb_file, removeHs = False)
preparator = MoleculePreparation()
preparator.prepare(m3d)
pdbqt_string = preparator.write_pdbqt_string()
pdbqt_string_ready = pdbqt_string.replace('UNL','MOL')
ligand_file = '/Users/amyhe/Downloads/test.pdbqt'
with open(ligand_file,'w') as pdbqt_file:
    pdbqt_file.write(pdbqt_string_ready)
```

You will have the following PDBQT file:

```
REMARK SMILES CCCCO
REMARK SMILES IDX 3 1 2 2 1 3 4 4 5 5
REMARK H PARENT 5 6
ROOT
ATOM      1  C3  MOL     1      -0.417   0.816   0.570  1.00  0.00     0.028 C 
ENDROOT
BRANCH   1   2
ATOM      2  C2  MOL     1       0.829   0.814  -0.287  1.00  0.00    -0.001 C 
ATOM      3  C1  MOL     1       1.508  -0.516  -0.314  1.00  0.00     0.004 C 
ENDBRANCH   1   2
BRANCH   1   4
ATOM      4  C4  MOL     1      -1.460  -0.154   0.122  1.00  0.00     0.156 C 
BRANCH   4   5
ATOM      5  O1  MOL     1      -1.015  -1.471   0.126  1.00  0.00    -0.396 OA
ATOM      6  H10 MOL     1      -1.759  -2.109   0.359  1.00  0.00     0.210 HD
ENDBRANCH   4   5
ENDBRANCH   1   4
TORSDOF 3
```

Note: It's possible (but optional) to have unique atom names in MOL2 files, as inputs for command line script mk_prepare_ligand.py. However, RDKit doesn't parse the atom names from MOL2 files. Therefore, currently there isn't a way to retain atom names from MOL2 files without a rewrite of the parser.

**Comment by diogomart** [comment]
@rwxayheee thank you for looking into this. I think generating unique names may be useful, so let's keep this issue open, but not a priority for v0.6.0, so I'll remove this issue from the v0.6.0 milestone.

**Comment by rwxayheee** [comment]
@diogomart Got it! Yes I agree generating unique names may be useful. I will continue working on this when I get a chance

**Comment by rwxayheee** [comment]
This will be addressed with #233. Just for record

**Comment by rwxayheee** [comment]
Included #255. Thanks again for bringing this to our attention, @alfredoq. Please feel free to reopen or leave a comment if there are any further issues
